File: go/socketd/transport/stream/impl/stream_base.go
package impl

import (
	"context"
	"fmt"
	"time"

	"socketd/transport/stream"
)

type StreamBase struct {
	sid             string
	demands         int
	timeout         time.Duration
	doOnError       func(err error)
	doOnProgress    func(isSend bool, val int, max int)
	insuranceCancel context.CancelFunc
}

func NewStreamBase(sid string, demands int, timeout time.Duration) *StreamBase {
	return &StreamBase{
		sid:     sid,
		demands: demands,
		timeout: timeout,
	}
}

func (b *StreamBase) Sid() string {
	return b.sid
}

func (b *StreamBase) Demands() int {
	return b.demands
}

func (b *StreamBase) Timeout() time.Duration {
	return b.timeout
}

func (b *StreamBase) SetTimeout() time.Duration {
	return b.timeout
}

func (b *StreamBase) InsuranceStart(streamManger stream.StreamManager, streamTimeout time.Duration) {
	if b.insuranceCancel != nil {
		return
	}

	var ctx, cancel = context.WithCancel(context.Background())
	b.insuranceCancel = cancel

	var timer = time.NewTimer(streamTimeout)
	go func(ctx context.Context) {
		select {
		case <-ctx.Done():
		case <-timer.C:
			streamManger.Remove(b.sid)
			b.OnError(fmt.Errorf("the stream response timeout,sid = %s", b.sid))
		}
	}(ctx)
}

func (b *StreamBase) InsuranceCancel() {
	if b.insuranceCancel != nil {
		b.insuranceCancel()
	}
}

func (b *StreamBase) OnError(err error) {
	if b.doOnError != nil {
		b.doOnError(err)
	}
}

func (b *StreamBase) OnProgress(isSend bool, val int, max int) {
	if b.doOnProgress != nil {
		b.doOnProgress(isSend, val, max)
	}
}

func (b *StreamBase) ThenError(onError func(err error)) {
	b.doOnError = onError
}

func (b *StreamBase) ThenProgress(onProgress func(isSend bool, val int, max int)) {
	b.doOnProgress = onProgress
}

File: swift/socketd/socketd/src/socketd/transport/stream/impl/StreamBase.swift
//
//  StreamBase.swift
//  socketd
//
//  Created by noear on 2024/1/12.
//

import Foundation

class StreamBase : StreamInternal{
    private var _sid:String!;
    private var _demands:Int32!;
    private var _timeout:Int64!;
    
    private var _doOnError: IoConsumer<Error>?;
    private var _doOnProgress: IoTriConsumer<Bool, Int32, Int32>?;
    
    init(_ sid:String, _ demands:Int32, _ timeout:Int64) {
        self._sid = sid;
        self._demands = demands;
        self._timeout = timeout;
    }
    
    
    func sid() -> String {
        return _sid;
    }
    
    func demands() -> Int32 {
        return _demands;
    }
    
    func timeout() -> Int64 {
        return _timeout;
    }
    
    func insuranceStart(_ streamManger: StreamManger, _ streamTimeout: Int64) {
        
    }
    
    func insuranceCancel() {
        
    }
    
    func onError(_ error: Error) {
        if(_doOnError != nil){
            _doOnError!(error);
        }
    }
    
    
    func thenError(_ onError: @escaping IoConsumer<Error>) -> Self {
        _doOnError = onError;
        
        return self;
    }
    
    func onProgress(_ isSend: Bool, _ val: Int32, _ max: Int32) {
        if(_doOnProgress != nil){
            _doOnProgress!(isSend, val, max);
        }
    }
    
    func thenProgress(_ onProgress: @escaping IoTriConsumer<Bool, Int32, Int32>) -> Self {
        _doOnProgress = onProgress;
        
        return self;
    }
    
    
    func onReply(_ reply: MessageInternal) {
        
    }
    
    
    func isDone() -> Bool {
        return false;
    }
}
